lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp

```diff
@@ -0,0 +1,427 @@
+        // if any of those values do not dominate the latch, they would need
+        // to be propagated to the latch, which we don't do right now.
+        //
+        if (Instruction *value = dyn_cast<Instruction>(phi->getIncomingValueForBlock(exiting_block))) {
+          if (!DT->dominates(value, latch)) {
+            return false;
+          }
+        }
+      }
+      else {
+        break;
+      }
+    }
+  }
+
+  BranchInst *exiting_br = cast<BranchInst>(exiting_block->getTerminator());
+  Value *exit_cond = exiting_br->getCondition();
+
+  Value *exit_cond_dominates_latch = nullptr;
+  SmallVector<std::pair<BasicBlock *, Value *>, 4> blocks_with_side_effect;
+  bool give_up = false;
+  std::unordered_map<BasicBlock *, PHINode *> cached_phis;
+
+  // Use a worklist to propagate the exit condition from within the block
+  {
+    Value *false_value = ConstantInt::getFalse(i1Ty);
+
+    struct Propagate_Data {
+      BasicBlock *bb;
+      Value *value;
+      bool is_exiting_block;
+    };
+
+    std::unordered_set<BasicBlock *> seen;
+    SmallVector<Propagate_Data, 4> work_list;
+
+    work_list.push_back({ exiting_block, exit_cond, true });
+    seen.insert(exiting_block);
+
+    BasicBlock *latch = L->getLoopLatch();
+
+    for (unsigned i = 0; i < work_list.size(); i++) {
+      Propagate_Data data = work_list[i];
+
+      BasicBlock *bb = data.bb;
+
+      // Do not include the exiting block itself in this calculation
+      if (!data.is_exiting_block) {
+        // If this block is part of an inner loop... Give up for now.
+        if (LI->getLoopFor(data.bb) != L) {
+          give_up = true;
+        }
+        // Otherwise just remember the blocks with side effects (including the latch)
+        else {
+          for (Instruction &I : *bb) {
+            if (I.mayWriteToMemory() && !IsNoop(&I)) {
+              blocks_with_side_effect.push_back({ bb, data.value });
+              break;
+            }
+          }
+        }
+      } // If this is not the first iteration
+
+      // Don't continue to propagate when we hit the latch
+      if (data.bb == latch /*|| DT->dominates(data.value, latch)*/) {
+        exit_cond_dominates_latch = data.value;
+        continue;
+      }
+
+      for (BasicBlock *succ : llvm::successors(bb)) {
+        if (!L->contains(succ))
+          continue;
+
+        PHINode *phi = cached_phis[succ];
+        if (!phi) {
+          phi = PHINode::Create(i1Ty, 2, "dx.struct_exit.exit_cond", &*succ->begin());
+          for (BasicBlock *pred : llvm::predecessors(succ)) {
+            phi->addIncoming(false_value, pred);
+          }
+          cached_phis[succ] = phi;
+        }
+
+        for (unsigned i = 0; i < phi->getNumIncomingValues(); i++) {
+          if (phi->getIncomingBlock(i) == bb) {
+            phi->setIncomingValue(i, data.value);
+            break;
+          }
+        }
+
+        if (!seen.count(succ)) {
+          work_list.push_back({ succ, phi });
```

Comment:
I've confirmed that changing this to `{ succ, phi, false }` is all that remains to get the Linux build working.

Reply:
thank you.